drivers/adc/adc_max11102_17.c

```diff
@@ -0,0 +1,454 @@
+				      bool wait)
+{
+	int result;
+	struct max11102_17_data *data = dev->data;
+
+	result = max11102_17_validate_sequence(dev, sequence);
+
+	if (result != 0) {
+		LOG_ERR("sequence validation failed");
+		return result;
+	}
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	if (wait) {
+		result = adc_context_wait_for_completion(&data->ctx);
+	}
+
+	return result;
+}
+
+static int max11102_17_read_sample(const struct device *dev, int16_t *sample)
+{
+	const struct max11102_17_config *config = dev->config;
+	int result;
+	size_t trailing_bits = 15 - config->resolution;
+	uint8_t buffer_rx[2];
+	const struct spi_buf rx_buf[] = {{
+		.buf = buffer_rx,
+		.len = ARRAY_SIZE(buffer_rx),
+	}};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+
+	result = spi_read_dt(&config->bus, &rx);
+
+	if (result != 0) {
+		LOG_ERR("read failed with error %i", result);
+		return result;
+	}
+
+	*sample = sys_get_be16(buffer_rx);
+	LOG_DBG("raw sample: 0x%04X", *sample);
+
+	*sample = *sample >> trailing_bits;
+	*sample = *sample & GENMASK(config->resolution, 0);
+	LOG_DBG("sample: 0x%04X", *sample);
+
+	return 0;
+}
+
+static int max11102_17_adc_perform_read(const struct device *dev)
+{
+	int result;
+	struct max11102_17_data *data = dev->data;
+
+	k_sem_take(&data->acquire_signal, K_FOREVER);
+
+	result = max11102_17_read_sample(dev, data->buffer);
+	if (result != 0) {
+		LOG_ERR("reading sample failed");
+		adc_context_complete(&data->ctx, result);
+		return result;
+	}
+
+	data->buffer++;
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+
+	return result;
+}
+
+#if CONFIG_ADC_ASYNC
+static int max11102_17_adc_read_async(const struct device *dev, const struct adc_sequence *sequence,
+				      struct k_poll_signal *async)
+{
+	int result;
+	struct max11102_17_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, true, async);
+	result = max11102_17_adc_start_read(dev, sequence, true);
+	adc_context_release(&data->ctx, result);
+
+	return result;
+}
+
+static int max11102_17_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	int result;
+	struct max11102_17_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	result = max11102_17_adc_start_read(dev, sequence, true);
+	adc_context_release(&data->ctx, result);
+
+	return result;
+}
+
+#else
+static int max11102_17_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	int result;
+	struct max11102_17_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	result = max11102_17_adc_start_read(dev, sequence, false);
+
+	while (result == 0 && k_sem_take(&data->ctx.sync, K_NO_WAIT) != 0) {
+		result = max11102_17_adc_perform_read(dev);
+	}
+
+	adc_context_release(&data->ctx, result);
+	return result;
+}
+#endif
+
+#if CONFIG_ADC_ASYNC
+static void max11102_17_acquisition_thread(struct device *dev)
```

Comment:
Done, BUT: Looking into this I realized that k_thread_entry_t is actually defined as (void)(void*, void*, void*). At this point I am not sure if it might fuck up the stack if the signature of the actual function and the function pointer through which the call happens do not match?
Anyway, this seems to be the common approach, so at least it is now consistently dangerous ;-).